python/cudf/dataframe/multiindex.py

```diff
@@ -0,0 +1,341 @@
+        # ---------------
+        if len(row_tuple) < len(self.levels):
+            columns = self._popn(len(row_tuple))
+            result.columns = columns.take(valid_indices)
+        else:
+            result.columns = self.take(valid_indices)
+        if len(result.columns.levels) == 1:
+            columns = []
+            for code in result.columns.codes[result.columns.codes.columns[0]]:
+                columns.append(result.columns.levels[0][code])
+            name = result.columns.names[0]
+            result.columns = StringIndex(columns, name=name)
+        return result
+
+    def __len__(self):
+        return len(self.codes[self.codes.columns[0]])
+
+    def __eq__(self, other):
+        return self.levels == other.levels and\
+                self.codes == other.codes and\
+                self.names == other.names
+
+    @property
+    def is_contiguous(self):
+        return True
+
+    @property
+    def size(self):
+        return len(self.codes[0])
+
+    def take(self, indices):
+        from collections.abc import Sequence
+        from cudf import Series
+        from numbers import Integral
+        if isinstance(indices, (Integral, Sequence)):
+            indices = np.array(indices)
+        elif isinstance(indices, Series):
+            indices = indices.to_gpu_array()
+        codes = self.codes.take(indices)
+        result = MultiIndex(self.levels, codes)
+        result.names = self.names
+        return result
+
+    def __iter__(self):
+        self.n = 0
+        return self
+
+    def __next__(self):
+        if self.n < len(self.codes):
+            result = self[self.n]
+            self.n += 1
+            return result
+        else:
+            raise StopIteration
+
+    def __getitem__(self, index):
+        match = self.take(index)
+        result = []
+        for level, item in enumerate(match.codes):
+            result.append(match.levels[level][match.codes[item][0]])
+        return tuple(result)
+
+    """
+    def __getitem__(self, index):
+        if isinstance(index, slice):
+            # a new MultiIndex with the sliced codes, same levels and names
+            None
+        elif isinstance(index, int):
+            # a tuple of the labels of the item defined by the set of codes
+            # at the ith positionjA
+            None
+        elif isinstance(index, (list, np.ndarray)):
+            result = MultiIndex(self.levels, self.codes.iloc[index])
+            result.names = self.names
+            return result
+        if isinstance(index, (DeviceNDArray)):
+            return self.take(index)
+        else:
+            raise IndexError('only integers, slices (`:`), ellipsis (`...`),'
+            'numpy.newaxis (`None`) and integer or boolean arrays are valid '
+            'indices')  # noqa: E128
+    """
+
+    @property
+    def _values(self):
+        return list([i for i in self])
+
+    @classmethod
+    def from_tuples(cls, tuples, names=None):
+        # cheating
```

Comment:
Not cheating, just not reinventing the wheel! 😄 
```suggestion
        # Use Pandas for handling Python host objects
```